### PR TITLE
Databases: Use LibGuides alternate names field

### PR DESCRIPTION
**Pull request by maccabeelevine**
LibGuides AZ has an "alternate names" field described as

> Alternate names associated with the database. If a user searches for these words, this database will appear in the results.

If the LibGuides option is enabled for the Databases recommendation module, we should (by default) search for URLs against the alternate names in addition to the primary name.

An example we ran into locally is Oxford Music Online often referred to as Grove Music Online by long-time researchers.  So both "Oxford" and "Grove" should match that database URL.

### CODE REVIEW

File: module/VuFind/src/VuFind/Recommend/Databases.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Recommend/Databases.php:335]: I notice that the API field is plural (alt_names) but we're treating it as a single flat string. What happens when there are multiple values here? Is there a standard delimiter, or is this just a free-text field that's used inconsistently?

I realize that we're doing a str_contains search against this value, so it should work even if the field contains multiple names... but if there is a standard and reliable delimiter, it feels like it might be worth splitting it up to make our index a little more accurate -- that might eliminate a certain rare subset of false positive hits and/or provide an opportunity for smarter relevance ranking logic in the future.

**Review by demiankatz — COMMENTED** [review]
I don't have a problem with merging this as-is, but I wanted to raise one point before doing it in case there's room for improvement here.

Even if we change nothing, maybe it's worth adding a comment to address the point raised, so I don't forget and ask it again in the future. ;-)

File: module/VuFind/src/VuFind/Recommend/Databases.php
Comment (maccabeelevine) [reply module/VuFind/src/VuFind/Recommend/Databases.php:335]: Yeah, unfortunately it is a plain single-line string field.  I have seen a value like "Used to be called XX" which is probably not what was intended and of course would cause problems if someone actually searched for "used to" for some ^$E@#%! reason.  So I went with a naive algorithm.  In any case that's what LibGuides would be doing itself to search against that field.

Comment (demiankatz) [reply module/VuFind/src/VuFind/Recommend/Databases.php:335]: Makes sense! As I noted, it's then probably worth adding a comment like:

```
// The alt_names field is single-valued free text; we'll do the best we can with it
```

:-)

File: module/VuFind/src/VuFind/Recommend/Databases.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Recommend/Databases.php]: Since this property isn't set until `setConfig` is called, it might be a good idea to give it a default value. Of course, `setConfig` should always be called in normal circumstances, but it might be good to avoid errors if the class is used in a different way in the future. (Same probably goes for a couple of other existing properties).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the clarification, @maccabeelevine -- just a couple of small comments.

File: module/VuFind/src/VuFind/Recommend/Databases.php
Comment (maccabeelevine) [reply module/VuFind/src/VuFind/Recommend/Databases.php:335]: Agreed, sorry I missed the comment suggestion the first time around.

File: module/VuFind/src/VuFind/Recommend/Databases.php
Comment (maccabeelevine) [reply module/VuFind/src/VuFind/Recommend/Databases.php]: I agree in theory with the point about protecting if setConfig is not called in the future, but I don't like the idea of declaring the same default value (for each property) in two different places.  That seems to create a maintenance issue  if the default changes or we deprecate something, and it makes the code less clear.  Of course declaring a different default value in the property declaration is no better.  The best approach might be declare defaults and then to change setConfig() to only set the values if configured, i.e.  

if (isset($databasesConfig->useQuery) { $this->useQuery = $databasesConfig->useQuery; }

which is verbose, or

$this->useQuery = $databasesConfig->useQuery ?? $this->useQuery;

which is shorter but looks silly.  And neither seems to be the norm in this codebase generally, or in recommendation modules in particular; i.e. no defaults set during property declaration in AbstractSearchObject.  So I agree with the future-proofing problem but I think the alternatives are worse, unless we want to start a much larger conversation?

Comment (demiankatz) [reply module/VuFind/src/VuFind/Recommend/Databases.php]: I actually wouldn't mind either of these approaches. As you say, it may also be a good idea to check other Recommendation modules for problems that fit this pattern (it's an easy thing to overlook, and probably has been overlooked elsewhere), but I don't think we have to solve the problem everywhere in order to address it here. Both of those patterns can be found elsewhere in the codebase (particularly in the Search\*\Options classes), though you might not find such precedents in the Recommendation code, as it is generally not very uniform, since each module is designed to be fairly stand-alone, and many come from different developers.

There's also a third potential approach: create a getter for the property, and have it throw an exception if the getter is called before the value has been set.

I think of the three options, I slightly prefer the null coalescing approach because of its brevity, in spite of the fact that as you say, it looks a bit silly. But I could live with any of them!

Comment (maccabeelevine) [reply module/VuFind/src/VuFind/Recommend/Databases.php]: Ok, as you say I did find examples of both approaches I had suggested, even if they look a bit odd.  I went with the null coalescing approach, like you I think it's best.  As long as there is a comment explaining it, which I adapted from Base/Options.

**Review by demiankatz — APPROVED** [review]
Thanks, @maccabeelevine!

One small point that I'll leave to your discretion. With new syntax introduced in PHP 8 (and thus not yet used widely in the codebase), I'm pretty sure you can simplify:

```
        $this->configFileDatabases = isset($databasesConfig->url) ? $databasesConfig->url->toArray()
            : $this->configFileDatabases;
```

to:

```
        $this->configFileDatabases = $databasesConfig?->url?->toArray() ?? $this->configFileDatabases;
```

I've approved this PR but won't merge it right away, so I can leave this up to you: if you want to try that simplification and it works, feel free to make the change and merge (I trust you!). If you've spent enough time on this and don't want to bother, then just merge as-is and we can work on applying that optimization more globally in the future.

**Comment by maccabeelevine** [comment]
> new syntax introduced in PHP 8

@demiankatz Nice!  I am fan of the same in javascript, glad to see it is in PHP now as well.  I made those changes.  To be clear I did not also null-safe the $databasesConfig variable since that is checked earlier on line 172.

**Pull request merged**
